presto-spi/src/main/java/com/facebook/presto/spi/relation/RowExpressionMatcher.java

```diff
@@ -0,0 +1,120 @@
+package com.facebook.presto.spi.relation;
+
+import com.facebook.presto.spi.block.AbstractRowBlock;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.RowBlock;
+import com.facebook.presto.spi.type.ArrayType;
+import com.facebook.presto.spi.type.RowType;
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.SliceOutput;
+
+import java.util.Iterator;
+import java.util.List;
+
+import static java.lang.Math.toIntExact;
+
+public class RowExpressionMatcher
+        implements RowExpressionVisitor<Boolean, RowExpression>
+{
+
+    @Override
+    public Boolean visitCall(CallExpression call, RowExpression context)
+    {
+        //TODO Please Implement
+        return false;
+    }
+
+    @Override
+    public Boolean visitInputReference(InputReferenceExpression reference, RowExpression context)
+    {
+        return false;
+    }
+
+    @Override
+    public Boolean visitConstant(ConstantExpression actual, RowExpression expected)
+    {
+        if (!(expected instanceof ConstantExpression)) {
+            return false;
+        }
+
+        if (!expected.getType().equals(actual.getType())) {
+            return false;
+        }
+
+        Type type = actual.getType();
+
+        if (type instanceof RowType) {
+            RowType actualRowType = (RowType) actual.getType();
+            RowType expectedRowType = (RowType) expected.getType();
+
+            List<Type> actualTypes = actualRowType.getTypeParameters();
+            List<Type> expectedTypes = expectedRowType.getTypeParameters();
+
+            if (actualTypes.size() != expectedTypes.size()) {
+                return false;
+            }
+
+            Iterator<Type> actualTypeItr = actualTypes.iterator();
+            Iterator<Type> expectedITypetr = expectedTypes.iterator();
+
+            while (actualTypeItr.hasNext() && expectedITypetr.hasNext()) {
+                if (actualTypeItr.next() != expectedITypetr.next()) {
+                    return false;
+                }
+            }
+
+            Object actualObj = actual.getValue();
+            Object expectedObj = ((ConstantExpression) expected).getValue();
+            if (!(actualObj instanceof RowBlock) || !(expectedObj instanceof RowBlock)) {
```

Comment:
Let's move the object comparison part to a separate overridable method.

Reply:
And handle it similar to `LiteralEncoder`.

Reply:
We can imagine, we will want to call this in both `ExpressionEquivalence` or `ExpressionMatcher`